src/Raven.Server/Documents/Replication/ReplicationLoader.cs

```diff
@@ -180,50 +180,101 @@ public ReplicationLoader(DocumentDatabase database, ServerStore server)
         }
 
         public void AcceptIncomingConnection(TcpConnectionOptions tcpConnectionOptions,
-            TcpConnectionHeaderMessage.OperationTypes headerOperation,
+            TcpConnectionHeaderMessage header,
+            X509Certificate2 certificate,
             JsonOperationContext.MemoryBuffer buffer)
         {
             var supportedVersions =
                 TcpConnectionHeaderMessage.GetSupportedFeaturesFor(TcpConnectionHeaderMessage.OperationTypes.Replication, tcpConnectionOptions.ProtocolVersion);
+            
+            string[] allowedPaths = null;
+            string pullDefinitionName = null;
 
-            if (supportedVersions.Replication.PullReplication)
+            switch (header.AuthorizeInfo?.AuthorizeAs)
             {
-                // wait for replication type
-                using (tcpConnectionOptions.ContextPool.AllocateOperationContext(out JsonOperationContext context))
-                using (var readerObject = context.ParseToMemory(
-                    tcpConnectionOptions.Stream,
-                    "initial-replication-message",
-                    BlittableJsonDocumentBuilder.UsageMode.None,
-                    buffer))
-                {
-                    var initialRequest = JsonDeserializationServer.ReplicationInitialRequest(readerObject);
-                    if (initialRequest.PullReplicationDefinitionName != null)
+                case TcpConnectionHeaderMessage.AuthorizationInfo.AuthorizeMethod.PullReplication:
+                case TcpConnectionHeaderMessage.AuthorizationInfo.AuthorizeMethod.PushReplication:
+                    if (supportedVersions.Replication.PullReplication == false)
```

Comment:
Shouldn't this `if` come before the `switch` ?

Reply:
It is a new feature so we also need to add to the supported versions an option for push/filtered replication.

Reply:
No, it shouldn't. We need to do this check if the other side requested pull replication in this manner. 
There should be no way that you'll get here if you aren't supporting pull replication